### ISSUE TITLE
[inductor] Debug the perf problem for pytorch_stargan

### ISSUE THREAD

**Issue by desertfire** [pytorch/torchdynamo#515]
The inference perf is 0.64x vs. 1.58x from NVFuser

**Comment by desertfire** [comment]
After running `nsys prof`, the following line stood out in the profiling report,

```
Time	Total Time	Instances	Avg	Med	Min	Max	StdDev	Category	Operation
88.6%	1.127 s	2825	398.843 μs	374.015 μs	373.470 μs	485.438 μs	39.316 μs	CUDA_KERNEL	void at::native::vectorized_elementwise_kernel<(int)4, at::native::FillFunctor<signed char>, at::detail::Array<char *, (int)1>>(int, T2, T3)
```

**Comment by desertfire** [comment]
Hmm, I use the same command line. What's your CUDA version? I use 11.6

**Comment by desertfire** [comment]
https://github.com/pytorch/torchdynamo/pull/518 doesn't provide much perf gain. I think `FillFunctor` is still the main problem. @ngimel , I would appreciate your insights on which ops could introduce those.

**Comment by ngimel** [comment]
P514136351 - the profile for the inductor-optimized workload doesn't show any fill kernel, but for some reason even with cuda graphs there's a lot of white space on the GPU timeline, I don't really understand why

**Comment by ngimel** [comment]
Yep, cuda graphs are not used due to mutated inputs, so, even though inductor compiles reasonable kernels whereas it looks like ` --backend aot_autograd` doesn't pick up any fusions at all, the resulting time for inductor is larger due to poor gpu utilization + python overhead (aot autograd goes to torchscript and thus is faster purely due to python overhead removal). 
@desertfire do you see any kernels with CodeGen in the name in the `--nvfuser --backend aot_autograd` (I presume that's how you run nvfuser)? I don't, but my observed nvfuser speed-up is also smaller.

**Comment by jansel** [comment]
> I also use 11.6, on AWS A100 Edit: I was on @jansel's inductor branch, torchdynamo head produces correct result. cc @jansel

A hrm, will need to bisect this one.  My first guess is I set the config for max_tiles to 2 in my branch, which may not play nice with some indexing dependency changes.

**Comment by desertfire** [comment]
> Yep, cuda graphs are not used due to mutated inputs, so, even though inductor compiles reasonable kernels whereas it looks like ` --backend aot_autograd` doesn't pick up any fusions at all, the resulting time for inductor is larger due to poor gpu utilization + python overhead (aot autograd goes to torchscript and thus is faster purely due to python overhead removal). @desertfire do you see any kernels with CodeGen in the name in the `--nvfuser --backend aot_autograd` (I presume that's how you run nvfuser)? I don't, but my observed nvfuser speed-up is also smaller.

I was looking at inference only, since this is the last torchbench test that we show a slowdown instead of improvement. This is the command line that I got from Jason,`./benchmarks/torchbench.py --cosine -dcuda --float16 --isolate -n50 --backend=cudagraphs_ts --nvfuser -k pytorch_stargan`.

**Comment by desertfire** [comment]
Discussed with @ngimel offline, the gaps we are seeing between kernels are likely because of Triton kernel launch overhead. If I comment out https://github.com/pytorch/torchdynamo/blob/5f2f374d9bbc6374fe725a17182e7d4c270c6833/torchinductor/compile_fx.py#L143, the perf will become 1.4x.

Given torchinductor is working in the framework of torchdynamo, is there any way we can relax this input mutation checking when combining with certain GUARDs?

**Comment by ngimel** [comment]
@jansel where's the code that avoids copying parameters to cuda graphs static inputs? As you've suggested we should cross-reference mutated inputs against not-copied list to make cuda graphs more permissive

**Comment by jansel** [comment]
For training the non-copied params are computed in these two places:
https://github.com/pytorch/torchdynamo/blob/39302c40a24ab574679a52655dc3d56dab59bc8c/torchinductor/compile_fx.py#L248
https://github.com/pytorch/torchdynamo/blob/39302c40a24ab574679a52655dc3d56dab59bc8c/torchinductor/compile_fx.py#L244

In each of those cases, the first N params are fixed and not copied.


For inference, the non-copied params are added in `wrap()` (see `*params_flat`), which is hidden from cudagraphs:  
https://github.com/pytorch/torchdynamo/blob/39302c40a24ab574679a52655dc3d56dab59bc8c/torchinductor/compile_fx.py#L244

Basically the `len(example_inputs)` will be smaller than the inputs to the FX graph.  If you count the graph placeholders, the first `N-len(example_inputs)` are not copied.

**Comment by desertfire** [comment]
The main problem here is how to reduce kernel launch overhead, which is being tracked at https://github.com/pytorch/torchdynamo/issues/746. Let's close this one.
